pkg/webconfig/tls_credentials.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webconfig
+
+import (
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/assets"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"path"
+)
+
+var (
+	volumePrefix = "web-config-tls-"
+)
+
+// tlsCredentials are the credentials used for web TLS.
+type tlsCredentials struct {
+	// mountPath is the directory where TLS credentials are intended to be mounted
+	mountPath string
+
+	// keySecret is the kubernetes secret containing the TLS key
+	keySecret v1.SecretKeySelector
+	// cert is the kubernetes secret or configmap containing the TLS certificate
+	cert monitoringv1.SecretOrConfigMap
+	// clientCA is the kubernetes secret or configmap containing the client CA certificate
+	clientCA monitoringv1.SecretOrConfigMap
+}
+
+// newTLSCredentials creates new tlsCredentials from secrets of configmaps.
+func newTLSCredentials(
+	mountPath string,
+	key v1.SecretKeySelector,
+	cert monitoringv1.SecretOrConfigMap,
+	clientCA monitoringv1.SecretOrConfigMap,
+) *tlsCredentials {
+	return &tlsCredentials{
+		mountPath: mountPath,
+		keySecret: key,
+		cert:      cert,
+		clientCA:  clientCA,
+	}
+}
+
+// getKeyMountPath is the mount path of the TLS key inside a prometheus container.
+func (a tlsCredentials) getKeyMountPath() string {
+	secret := monitoringv1.SecretOrConfigMap{Secret: &a.keySecret}
+	return a.assetPath(secret)
+}
+
+// getCertMountPath is the mount path of the TLS certificate inside a prometheus container,
+func (a tlsCredentials) getCertMountPath() string {
+	if a.cert.ConfigMap != nil || a.cert.Secret != nil {
+		return a.assetPath(a.cert)
+	}
+
+	return ""
+}
+
+// getCAMountPath is the mount path of the client CA certificate inside a prometheus container.
+func (a tlsCredentials) getCAMountPath() string {
+	if a.clientCA.ConfigMap != nil || a.clientCA.Secret != nil {
+		return a.assetPath(a.clientCA)
+	}
+
+	return ""
+}
+
+func (a tlsCredentials) assetPath(sel monitoringv1.SecretOrConfigMap) string {
+	return path.Join(a.mountPath, assets.TLSAssetKeyFromSelector("", sel).String())
```

Comment:
Here we are reusing the functionality from the assets package to generate a path inside the container. I am not sure if this is a good idea though, as changing the assets package will affect this one as well. I am open to any suggestions.